emulator: Implement Conditional Mutations

marete opened the pull request:
* Implement Conditional Mutations (CheckAndMutate)
* Implement tests covering all new code

<!-- Reviewable:start -->
- - -
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/Unoperate/google-cloud-cpp/4)
<!-- Reviewable:end -->

marete committed:
emulator: Start RowTransaction class.

Logic from Table::MutateRow() will be moved here and this commit
proposes the general structure of this class, for early review.

marete committed:
emulator: Implement undo log and wire-up rewind for the SetCell mutation.

- Introduces a RowTransaction class to proxy mutations and record undo
  logs in case we needs to rewind.
- Re-implements the SetCell mutation in the RowTransaction class.
- Introduces all Undo types/actions for SetCell and records them in an
  undo stack.
- For SetCell only and to facilitate testing with only SetCell
  mutations, introduces the use of RowTransaction in doing the mutation.
- clang-format is run on all affected files (which has removed some
  headers, e.t.c.

A test for SetCell is coming up and is not currently included here.

marete committed:
emulator: Properly declate all possible row mutations.

And unconditionally return an error status for the aggregation
mutations, which we don't currently support.

marete committed:
emulator: Implement a SetCell that records an undo log.

marete committed:
Complete implementation of Undo() for SetCell mutations.

marete committed:
emulator: Implement a constructor for the RowTransaction class.

marete committed:
emulator: Pass RowTransaction a constant reference to the mutation request.

marete committed:
emulator: Introduce the use of a RowTransaction in Table::MutateRow.

Currently in use only for SetCell.

marete committed:
emulator: Run clang-format on the subset of the files we have been changing.

marete committed:
Merge branch 'emulator' of github.com:marete/google-cloud-cpp into emulator.

marete committed:
emulator: Start test suite for row mutation transaction rollback.

marete committed:
emulator: Add a static method to Table that returns a default-constructed table.

marete committed:
Merge remote-tracking branch 'upstream/emulator' into emulator

marete committed:
emulator: testing: Create a table in the correct way.

Populate a schema protobuf with the minimal fields and call
Table::Create();

marete committed:
emulator: testing: rollback_test: Fix crash when setting up SetCell mutation.

There is a crash later on, in the code that sets up the undo log, that
will be fixed in an upcoming commit.

marete committed:
emulator: SetCell mutation: Various logic fixes for undo log.

marete committed:
emulator: transaction rollback test: Fix crash in destructor.

Use the correct sub-message allocation method, otherwise the protobuf
code becomes confused about who is the owner of the SetCell mutation
message and tries to free it twice, I think.

See also: https://protobuf.dev/reference/cpp/arenas/

marete committed:
emulator: trasnaction rollback test: Fix crash in destructor.

Use the correct sub-message allocation method, otherwise the protobuf
code becomes confused about who is the owner of the SetCell mutation
message and tries to free it twice, I think.

See also: https://protobuf.dev/reference/cpp/arenas/

marete committed:
Merge branch 'emulator' of github.com:marete/google-cloud-cpp into emulator

marete committed:
emulator: table: Add a method to return an iterator to the column families map.

marete committed:
emulator: Introduce methods to retrieve begin and end of column family map iteration.

marete committed:
emulator: Complete implementing test for basic transaction functionality.

The test ensures that SetCell mutations still work via the
RowTransaction class.

Rollback will be tested in an upcoming change.

marete committed:
emulator: rollback test: Fix for more consistent handling of the table shared pointer.

marete committed:
emulator: Store a std::shared_ptr<Table> in RowTransaction.

This is a more correct way of doing it than storing a raw pointer.

marete committed:
emulator: rollback tests: fix test for basic functionality.

Thanks to Marek for the hint on the major part of the fix, which
involved copying a reference instead of a value!

marete committed:
emulator: rollback tests: fix test for basic functionality.

Thanks to Marek for the hint on the major part of the fix, which
involved copying a reference instead of a value!

marete committed:
emulator: ColumnFamily class: Disable copying.

marete committed:
emulator: Disable copying ColumnFamilyRow and simplify tx SetCell logic.

marete committed:
Merge branch 'emulator' of github.com:marete/google-cloud-cpp into emulator

marete committed:
emulator: Disable the copying of ColumnRows.

marete committed:
emulator: rollback_test.cc: factor out helper to check a cell exists with a specific value.

marete committed:
emulator: add helper to set a cell to a value.

Also, run clang-format on the file.

marete committed:
emulator: rollback tests: implement helper function to create tables.

marete committed:
emulator: rollack tests: Fix comment describing test.

marete committed:
emulator: rollback tests: Implement additional testing helper functions.

- To check if a row exists
- To check if a column exists

marete committed:
emulator: add helper to test > 1 SetCell mutations at once.

marete committed:
emulator: use set_cells() to set possibly more than one SetCell mutation.

marete committed:
emualtor: transaction rollback: Implement working test for RestoreValue.

This also involves a fix to the code due to a bug revealed by the
test!

marete committed:
emulator: rollback_tests.cc: run clang-format.

marete committed:
emulator: transaction rollback: Avoid copying the possibly very large cell values.

marete committed:
emulator: transaction rollback: SetCell: Implement working test for DeleteValue.

marete committed:
emulator: transaction rollback: Implement working DeleteRow test.

marete committed:
emulator: transaction rollback: Implement a DeleteFromRow that updates the undo log.

Also, add code coverage output artefacts to ignored files.

marete committed:
emulator: transaction rollback: Implement undo for row deletions (DeleteFromFamily).

And be careful not to potentially very large values (use std::move on
the column qualifier and cell values.

marete committed:
emulator: RestoreRow undo log: Rename struct members more conventionally.

marete committed:
emulator: DeleteFromFamily mutation: Implement with rollback support using RowTransaction.

marete committed:
emulator: transaction rollback: Test basic functionality of a DeleteFromFamily which uses RowTransaction.

marete committed:
emulator: test rollback of a row deletion due to DeleteFromFamily.

And fix a code bug discovered by this test.

marete committed:
Merge remote-tracking branch 'upstream/emulator' into emulator

marete committed:
emulator: Fixes for review: Improve name of return varible.

In particular, make it clear that we are only examining the status,
and will not use the possibly valid status value (returned object).

marete committed:
emulator: Rename RestoreRow -> RestoreColumnFamilyRow.

marete committed:
emulator: style: Remove trailing underscore from struct member names.

marete committed:
emulator: Rename several iterator variables more appropriately.

marete committed:
emulator: Use std::move() to avoid a copy.

marete committed:
emulator: Rename another iterator properly.

marete committed:
emulator: Use std::move() to avoid another copy.

marete committed:
emulator: check and return from exceptional conditions first.

This is to remain in line with the rest of the code.

marete committed:
emualtor: Fix the naming of another iterator variable.

marete committed:
emulator: remove use of C++17 syntax feature (if with initializer).

We need to have the code compilable by a C++14 compiler.

marete committed:
emulator: rename another iterator variable more appropriately.

marete committed:
emulator: Add a comment justifying storing the column family iterator in the undo log.

In general, it is unsafe to store other kind of map iterators in the
undo log, since a subsequent mutation on the same row could delete the
underlying map, and invalidate the iterator. However, deleting a
column family is a schema operation that cannot happen while we hold
the table lock, therefore for column families only, it is safe to
store the iterators in the undo log.

marete committed:
emulator: transaction rollback: RestoreValue: Eliminate the storage of iterators in undo log.

marete committed:
emulator: actually remove iterator field from RestoreValue.

marete committed:
emulator: Correct and radically simplify SetCell undo code logic.

- Don't store iterators in the DeleteValue undo type (they may be
  invalidated by a subsequent mutation).

- Maintain the invariants that there is never an empty column or empty
  row by using the ColumnFamily::DeleteColumn() function to delete
  ranges of cells during undo. The latter takes care of deleting empty
  columns and empty rows, already.

- And therefore we can get rid of the DeleteRow and DeleteColumn undo
  types and associated code.

All tests continue to pass.

marete committed:
emulator: remove undo types that are no longer needed.

marete committed:
emulator: Delete definitions of rollback types no longer used.

marete committed:
emulator: return unimplemented error status for unsupported aggregation mutations.

Replaces FIXMEs which would silently fail when this mutations are attempted.

marete committed:
emulator: row mutations: use a const reference rather than copying the mutation.

marete committed:
emulator: Modify the Delete* methods to return vector of deleted cells.

And also use std::move() to avoid copying potentially large cell values.

These change is needed to simplify the logic of preparing undo logs
for the Delete* mutations. In particular, we don't have to re-iterate
the table.

marete committed:
emualator: Implement DeleteFromColumn mutation using RowTransaction.

marete committed:
emulator: Test transaction DeleteFromColumn for basic functionality.

marete committed:
emulator: DeleteFromColumn basic functionality test: test that the column is actually deleted.

marete committed:
emulator: make it possible to test a vector of DeleteFromColumn row mutations.

This is needed to test DeleteFromColumn rollback.

marete committed:
emulator: Test transaction rollback for DeleteFromColumn.

marete committed:
emulator: TEST(TransactonRollback, DeleteFromColumnRollback): also check that the cells are properly restored.

marete committed:
emulator: Prepare to support transaction rollbacks for DeleteRow mutation.

In particular, reutrn a map of vectors containing the column qualifers
to columns (cells) deleted, for the undo log.

marete committed:
emulator: DeleteRow: don't forget to erase the row key.

marete committed:
emulator: Fix segmentation fault caused by invalidated iterator in DeleteRow.

Also implement DeleteFromFamily in terms of the new DeleteRow which
returns RestoreValue vectors (and therefore eliminiate the use of one
more undo log type).

marete committed:
emulator: Fix segmentation fault caused by invalidated iterator in DeleteRow.

marete committed:
Merge branch 'emulator' of github.com:marete/google-cloud-cpp into emulator

marete committed:
emulator: clang-format.

marete committed:
emulator: transaction rollback: get rid of another undo log type.

This leaves us with just 2 undo log types: RestoreValue and
DeleteValue, a situation that is much simpler than before.

marete committed:
emulator: Implement a DeleteFromRow that supports rollback.

marete committed:
emulator: Test DeleteFromRow with rollback support and use it in MutateRow code.

marete committed:
emulator: Update and correct a comment.

marete committed:
Merge remote-tracking branch 'upstream/emulator' into emulator

marete committed:
erge remote-tracking branch 'upstream/emulator' into emulator

marete committed:
Merge branch 'emulator' of github.com:marete/google-cloud-cpp into emulator

marete committed:
emualtor: Remove now-unused undo type.

marete committed:
Merge remote-tracking branch 'upstream/emulator' into emulator

marete committed:
emulator: Fixes to avoid copying values that are possibly very large (cell contents).

As well as other fixes for review.

marete committed:
emulator: mutation transactions: use std::chrono:duration_cast.

Makes it easier and safer to change the resolution at a later date.

marete committed:
emulator: use invariant that rows are never empty to eliminate some code.

Preceding code erases all columns. The last column erase in
DeleteColumn will cause the row to be erased. So we don't have to
search for it again in the map and attempt to erase it.

marete committed:
emulator: optimization: make it possible to delete a column via a row iterator.

In this and other situations, this saves us a duplicate call to find
the row by saving and passing an iterator.

marete committed:
emulator: eliminate redundant storage of row_key in undo log.

The entire undo is in the context of a row, therefore all the methods
have access to a row key.

This also reduces the size of the undo log.

marete committed:
emulator: Use google-standard status builder everywhere in main code.

marete committed:
emulator: Add a FIXME to ask to use google-standard Status builder everywhere in test file.

marete committed:
emulator: Remove unused local var.

marete committed:
emulator: eliminate a C++-17ism.

marete committed:
emulator: DeleteRow: fix hang/crash due to use of invalidated row iterator.

marete committed:
emulator: remove non-Google-standard Status builders.

marete committed:
emulator: remove some outdated FIXMEs.

- We now have atomic rollback support for all implemented mutations.

marete committed:
emulator: Implement Conditional Mutations (CheckAndMutateRow).

* Implement Conditional Mutations (CheckAndMutate)
* Implement tests covering all new code

marete commented:
Implemented on top of my emulator branch, please review Conditional Mutations (CheckAndMutate), with passing tests. Thanks!

marete commented:
After fixes for the locking and making the function available to GRPC clients of the server, this PR is now ready for a review. PTAL.

marete committed:
emulator: make some std::chrono usage more type-safe.

marete committed:
emulator: std::move() to avoid another copy of a cell.

marete committed:
emulator: eliminate some code duplication.

marete committed:
emulator: SetCell: return overwritten value (if any).

This prepares the way to simplify the code that prepares the undo log,
and in particular it obviates the need to check to see if the
key (timestamp) existed before the SetCell mutation.

marete committed:
emulator: radically simplify undo log preparation code.

This is thanks to simply returning the old value (if any) rather than
performing almost endless iterations in this function! :)

marete committed:
emulator: ColumnFamily::DeleteRow(): radically simplify the function.

And also eliminate the risk of accessing an invalidated iterator.

marete committed:
Merge remote-tracking branch 'upstream/emulator' into emulator

marete committed:
Merge remote-tracking branch 'upstream/emulator' into emulator

marete commented:
I have also rebased this PR/branch on top of the large PR so that to review only those changes, just choose "Changes from 1 commit" which commit is 9ff9e14e704a42519. PTAL. Thanks.

dopiera commented on google/cloud/bigtable/emulator/table.cc:
```diff
@@ -270,6 +279,8 @@ Status Table::MutateRow(google::bigtable::v2::MutateRowRequest const& request) {
 
   return Status();
 }
+
```
Please avoid unrelated changes.

marete replied:
Done.
<!-- Reviewable comment -OOav8krDp71tSY6gdYr:b-896fix -->
<!-- Sent from Reviewable.io -->

dopiera commented on google/cloud/bigtable/emulator/server.cc:
```diff
@@ -74,8 +74,21 @@ class EmulatorService final : public btproto::Bigtable::Service {
+      return ToGrpcStatus(maybe_table.status());
+    }
+
+    auto status_or = (*maybe_table)->CheckAndMutateRow(*request);
```
Perhaps `maybe_response` would be more coherent with the rest?

marete replied:
Done.
<!-- Reviewable comment -OOZ5ZVj9DrcNM9mSYsC:b-896fix -->
<!-- Sent from Reviewable.io -->

dopiera commented on google/cloud/bigtable/emulator/server.cc:
```diff
@@ -74,8 +74,21 @@ class EmulatorService final : public btproto::Bigtable::Service {
+      return ToGrpcStatus(status_or.status());
+    }
+
+    auto res = status_or.value();
```
You're making two copies here.

You can avoid both of them by
`response = std::move(maybe_response).value();`

marete replied:
Done.
<!-- Reviewable comment -OOboWkS0nbEFt6zN8Kn:b-896fix -->
<!-- Sent from Reviewable.io -->

dopiera commented on google/cloud/bigtable/emulator/table.cc:
```diff
@@ -330,6 +341,91 @@ StatusOr<StringRangeSet> CreateStringRangeSet(
+                                      request.DebugString()));
+  }
+
+  google::bigtable::v2::RowSet row_set;
```
Rather than creating a ` google::bigtable::v2::RowSet`, which will be instantly discarded in `CreateStringRangeSet`, you can simply do:
```
auto range = std::shared_ptr<StringRangeSet();
range->Sum(StringRangeSet::Range(row_key, false, row_key, false)); 
```

marete replied:
Done.
<!-- Reviewable comment -OOcBuYI-e4z_ulkhlzC:b-896fix -->
<!-- Sent from Reviewable.io -->

dopiera commented on google/cloud/bigtable/emulator/table.cc:
```diff
@@ -330,6 +341,91 @@ StatusOr<StringRangeSet> CreateStringRangeSet(
+  auto range_set = std::make_shared<StringRangeSet>(*std::move(maybe_row_set));
+
+
+  auto table_stream_ctor = [range_set = std::move(range_set), this] {
```
Can you please extract this to a helper function so that we don't duplicate the code with `ReadRows`?

marete replied:
Done.
<!-- Reviewable comment -OObk7FzBGx-nV1cjifu:b-896fix -->
<!-- Sent from Reviewable.io -->

dopiera commented on google/cloud/bigtable/emulator/table.cc:
```diff
@@ -330,6 +341,91 @@ StatusOr<StringRangeSet> CreateStringRangeSet(
+  } else {
+    mutate_row_request.mutable_mutations()->Assign(
+        request.false_mutations().begin(), request.false_mutations().end());
+  }
```
Rather than constructing a `MutateRowRequest`, can you please extract create a helper function to be used by both this and `MutateRow()`, with a signature

`(std::string const& row_key, RepeatedPtrField< ::google::bigtable::v2::Mutation > const &mutations)`?

As a general rule, creating objects containing potentially more data just to fit the interface is IMO not a good idea. It reminds me of an old joke:
> A mathematician is given the task of hammering a nail into a piece of wood. He picks up a hammer and does it efficiently.
> Next, he's given a second piece of wood, this time with a nail already halfway in. He pulls the nail out, reduces the problem to the previous one, and hammers it in from scratch.

Seriously speaking, I think you should avoid this pattern, though because:
* this usually adds memory/computational cost to the operations (not least because of the need of repacking the arguments)
* the externally provided protos usually need to be validated because they come from external users, so you need to perform the validation multiple times
* you make it harder for yourself to express what you need from the input arguments (e.g. does the proto need to have the table name set?)
* it's harder to reason about the function just by looking at its parameters

Instead, please extract a common part of the function with the logic you need and use that.

marete replied:
Done.
<!-- Reviewable comment -OOdCHpW1MjQQo6EYcvK:b-896fix -->
<!-- Sent from Reviewable.io -->

marete committed:
emulator: SetCell: set timestamp to server time if passed 0 or negative timestamp.

marete committed:
emulator: transaction rollback: Check that a SetCell with timestamp 0 can be correctly rolled back.

marete committed:
emulator: docs: correct formatting of several comments in header.

marete committed:
emulator: tests: use a better EXPECT* macro in one place.

marete committed:
emulator: Implement ColumnFamily::DeleteTimeStamp().

This allows us to delete cells by specifying a specific timestamp
rather than a timestamp range which might otherwise involve possibly
non-type-safe calculations to determine the upper bound of the
timestamps (in particular doing the latter needs one to assume a
specific resolution for bigtable's timestamp storage).

marete committed:
emulator: transaction undo: radically simplify the code for DeleteValue.

marete committed:
emulator: rollback tests: perform some trivial refactoring.

marete committed:
emulator: conditional mutations: trivial changes for review.

marete committed:
emulator: create function implementing common code to create CellStreams.

marete committed:
emulator: conditional mutations: use CreateCellStream to eliminate code duplication.

marete committed:
emulator: ReadRows: eliminate some code duplication by using Table::CreateCellStream.

marete committed:
emulator: conditinal mutation RPC: eliminate 2 copies of the response.

marete committed:
emulator: conditional mutations: construct a StringRanegSet more directly.

In particular don't construct and immediately destroy a bigtable RowSet.

marete committed:
emulator: RowTransaction: replace request data member with row_key.

We only ever use request.row_key() when we use request in the
RowTransaction class. So we need only construct it with the row_key.

This paves the way for some refactoring that would allow us to
execute an array of mutations (with rollback on error) without first
creating a bigtable MutateRowRequest.

marete committed:
emulator: transactions: use new common logic to perform a sequence of mutations.

With possible rollback.

marete committed:
emulator: conditional mutations: eliminate creation and use of a MutationRowRequest.

marete reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/Unoperate/google-cloud-cpp/4)* status: 0 of 11 files reviewed, 6 unresolved discussions (waiting on @dopiera)



<!-- Sent from Reviewable.io -->

marete reviewed (commented):
Thanks for the careful review.

This is ready for another round of review. PTAL.

> *[Reviewable](https://reviewable.io/reviews/Unoperate/google-cloud-cpp/4#-:-OOdCMmaDvP_UXnIJcpU:b-8r7xv)* status: 0 of 11 files reviewed, 6 unresolved discussions (waiting on @dopiera)



<!-- Sent from Reviewable.io -->

marete committed:
emulator tests: Fix bigtable_emulator_range_set_test.

The tests aborted due to an assertion in the constructor of Range that
enforced the invariant that start <= end for all well-formed
Ranges. However in several tests testing the ends of ranges, start was
passed in as "unimportant" with end being "A" or "B". This violates
the invariant and causes an abort at runtime.

Fixes: TBL-62.

dopiera reviewed (changes requested):
> *[Reviewable](https://reviewable.io/reviews/Unoperate/google-cloud-cpp/4)* status: 0 of 11 files reviewed, 2 unresolved discussions

___
*[`google/cloud/bigtable/emulator/table.cc` line 297 at r9](https://reviewable.io/reviews/Unoperate/google-cloud-cpp/4#-OOx8QMd0dvC_GKYEWll:-OOx8QMd0dvC_GKYEWlm:b-g9qs9q) ([raw file](https://github.com/Unoperate/google-cloud-cpp/blob/eecfb1c22a5a0bd364702e18936cd956b9285181/google/cloud/bigtable/emulator/table.cc#L297)):*
> ```cpp
> StatusOr<CellStream> Table::CreateCellStream(
>     std::shared_ptr<StringRangeSet> range_set,
>     absl::optional<google::bigtable::v2::RowFilter> maybe_row_filter) const {
> ```

This signature implies making a copy of the row filter proto. Can you please make sure you don't? I think the simplest way to do it is to move the responsibility to apply filters if necessary to the caller.



_Code quote:_
```cpp
 absl::optional<google::bigtable::v2::RowFilter> maybe_row_filter
```
___
*[`google/cloud/bigtable/emulator/table.cc` line 608 at r9](https://reviewable.io/reviews/Unoperate/google-cloud-cpp/4#-OOx9P8gCWexQwKQJKVA:-OOx9P8gCWexQwKQJKVB:b-egau0h) ([raw file](https://github.com/Unoperate/google-cloud-cpp/blob/eecfb1c22a5a0bd364702e18936cd956b9285181/google/cloud/bigtable/emulator/table.cc#L608)):*
> ```cpp
>       std::chrono::microseconds(set_cell.timestamp_micros()));
> 
>   if (timestamp <= std::chrono::milliseconds::zero()) {
> ```

I think this doesn't belong to the `RowTransaction` class. IMO this class should only be responsible for maintaining the undo log. Can we move it to `Table`, please?


<!-- Sent from Reviewable.io -->

marete committed:
Merge remote-tracking branch 'upstream/emulator' into emulator

marete committed:
emulator: fixes for clang-tidy and clang-format.

Restricted only to changes I have made to this PR.

Tested (all tests pass).

marete committed:
Merge remote-tracking branch 'upstream/emulator' into emulator

marete committed:
Merge branch 'emulator' into fix-test-bigtable_emulator_range_set_test

marete committed:
emulator: move zero timestamp handling logic to the Table class.

marete committed:
emulator: CreateCellStream: avoid a copy of the filter if it is supplied.

marete reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/Unoperate/google-cloud-cpp/4)* status: 0 of 11 files reviewed, 2 unresolved discussions (waiting on @dopiera)

___
*[`google/cloud/bigtable/emulator/table.cc` line 297 at r9](https://reviewable.io/reviews/Unoperate/google-cloud-cpp/4#-OOx8QMd0dvC_GKYEWll:-OP5cYIK3XUv8_r7t5zS:bwmvfpn) ([raw file](https://github.com/Unoperate/google-cloud-cpp/blob/eecfb1c22a5a0bd364702e18936cd956b9285181/google/cloud/bigtable/emulator/table.cc#L297)):*
<details><summary><i>Previously, dopiera (Marek Dopiera) wrote…</i></summary><blockquote>

This signature implies making a copy of the row filter proto. Can you please make sure you don't? I think the simplest way to do it is to move the responsibility to apply filters if necessary to the caller.
</blockquote></details>

There are 2 call locations and I have used std::move() on both of them. RowFilter has a move constructor, so my understanding of what will happen is that it will be like calling std::optional::emplace(std::move(object)) which does no copying at all. Please correct me if I am wrong.
___
*[`google/cloud/bigtable/emulator/table.cc` line 608 at r9](https://reviewable.io/reviews/Unoperate/google-cloud-cpp/4#-OOx9P8gCWexQwKQJKVA:-OP5Zibp7eDn2azYUUTZ:b-896fix) ([raw file](https://github.com/Unoperate/google-cloud-cpp/blob/eecfb1c22a5a0bd364702e18936cd956b9285181/google/cloud/bigtable/emulator/table.cc#L608)):*
<details><summary><i>Previously, dopiera (Marek Dopiera) wrote…</i></summary><blockquote>

I think this doesn't belong to the `RowTransaction` class. IMO this class should only be responsible for maintaining the undo log. Can we move it to `Table`, please?
</blockquote></details>

Done.


<!-- Sent from Reviewable.io -->

marete commented:
PTAL.

dopiera commented on google/cloud/bigtable/emulator/table.cc:
```diff
@@ -220,14 +225,33 @@ StatusOr<std::reference_wrapper<ColumnFamily>> Table::FindColumnFamily(
-  assert(request.table_name() == schema_.name());
 
-  RowTransaction row_transaction(this->get(), request);
+  return MutateRowUnlocked(request);
```
Now that you have `DoMutationsWithPossibleRollback` I think you don't nees `MutateRowUnlocked` anymore.

marete replied:
Done.
<!-- Reviewable comment -OQ4tMtHA7XFm4mmRnca:b-896fix -->
<!-- Sent from Reviewable.io -->

dopiera commented on google/cloud/bigtable/emulator/table.h:
```diff
@@ -117,7 +132,9 @@ class RowTransaction {
   void commit() { committed_ = true; }
 
-  Status SetCell(::google::bigtable::v2::Mutation_SetCell const& set_cell);
+  Status SetCell(::google::bigtable::v2::Mutation_SetCell const& set_cell,
```
This function has become non-trivial - can you please add a comment on why `timestamp_override` is needed and what it does?

marete replied:
Done.
<!-- Reviewable comment -OQ4uFhV9EK6U3XfHetJ:b-896fix -->
<!-- Sent from Reviewable.io -->

dopiera commented on google/cloud/bigtable/emulator/table.h:
```diff
@@ -135,8 +152,9 @@ class RowTransaction {
-  ::google::bigtable::v2::MutateRowRequest const& request_;
+  std::stack<absl::variant<DeleteValue, RestoreValue>>
+      undo_;
+  std::string const& row_key_;
```
Can you please add a comment on why it is safe to keep the cref? I.e. that we don't want to copy a potentially large value and we rely on the fact that this is actually a cref from the request proto so it won't go away during the transaction?

marete replied:
Done.
<!-- Reviewable comment -OQ4uwT0BCWYKEcgpmAS:b-896fix -->
<!-- Sent from Reviewable.io -->

dopiera reviewed (approved):
Just some nits but otherwise LGTM. Feel free to merge without another round of review after you address them.

marete committed:
Merge remote-tracking branch 'upstream/emulator' into emulator

marete committed:
Merge branch 'emulator' into conditional_mutations

marete committed:
emulator: various fixes to implement review comments.

marete committed:
emulator: clang-format and other mechanical changes.

Changes made by:

./ci/cloudbuild/build.sh -t checkers-ci

marete reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/Unoperate/google-cloud-cpp/4)* status: 0 of 11 files reviewed, 5 unresolved discussions



<!-- Sent from Reviewable.io -->
